File: 15.3-sum.js
/*
 * @lc app=leetcode id=15 lang=javascript
 *
 * [15] 3Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  if (nums.length < 3) return [];

  if (nums.length === 3) {
    const sum = nums.reduce((a, b) => a + b, 0);
    if (sum === 0) return [nums];
    return [];
  }

  let result = [];

  // sort the array
  nums.sort((a, b) => a - b);

  for (let i = 0; i < nums.length - 2; i++) {
    // Already Checked this element
    if (i == 0 || (i > 0 && nums[i] == nums[i - 1])) continue;

    // ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~

    let lower = i - 1;
    let higher = nums.length - 1;

    // ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~

    while (lower < higher) {
      if (nums[i] + nums[lower] + nums[higher] === 0) {
        result.push([nums[i], nums[lower], nums[higher]]);

        // Prevenet duplicates higher and lowers
        while (lower < higher && nums[lower] + nums[lower + 1]) lower++;
        while (lower < higher && nums[higher] + nums[higher - 1]) higher--;

        lower++;
        higher--;
      } else if (nums[i] + nums[lower] + nums[higher] > 0) {
        higher--;
      } else {
        lower++;
      }
    }

    // ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
  }
  return result;
};
// @lc code=end

File: GO/15.3-sum.go
/*
 * @lc app=leetcode id=15 lang=golang
 *
 * [15] 3Sum
 */
package main

import "sort"

// @lc code=start
func threeSum(nums []int) [][]int {

	if len(nums) < 3 {
		return [][]int{}
	}

	sort.Ints(nums)
	res := [][]int{}

	for i := 0; i < len(nums)-2; i++ {

		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		front := i + 1
		rear := len(nums) - 1

		for front < rear {
			sum := nums[i] + nums[front] + nums[rear]
			if sum > 0 {
				rear--
			} else if sum < 0 {
				front++
			} else {
				res = append(res, []int{nums[i], nums[front], nums[rear]})
				front++
				for front < rear && nums[front] == nums[front-1] {
					front++
				}
			}
		}
	}
	return res
}

// @lc code=end
